### PR TITLE
Adding ARO SRE notification banner controller

### PR DESCRIPTION
**Pull request by Makdaam**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9396114/

### What this PR does / why we need it:

This PR allows ARO SRE to set a web console banner which asks the customer to contact Azure or Red Hat support. The wording of the message isn't set in stone yet. The code is ready for review.

Screenshot_2021-07-07_18-49-01

### Test plan for issue:

Start a dev cluster, edit the Cluster cluster object.
spec:
  features:
    reconcileBanner: true
  banner:
    content: 1
    
 the banner in Spec is a separate struct to include additional information via specific fields if we add more messages (ticket number, desired version, deadline date etc.).

### Is there any documentation that needs to be updated for this PR?

https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki/63811/Escalation-guidelines?anchor=%5Bdraft%2Fnot-on-prod-yet%5D - SOP proposal, I'll put it into a PowerShell script after it rolls out to production.

https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki/175282/Approved-Console-Notification-Messages - approved notification messages.

### CODE REVIEW

**Comment by nilsanderselde** [comment]
The wording sounds good. It's enough to capture their attention, and having a standard message makes it easy to flip on/off. The code changes are very straightforward, too. Nice job.

File: pkg/operator/controllers/banner/banner.go
Comment (nilsanderselde) [inline pkg/operator/controllers/banner/banner.go:41]: I would write "resourceid" as "resource ID"

Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go]: changed to resourceID

File: pkg/operator/deploy/staticresources/aro.openshift.io_clusters.yaml
Comment (gvanderpotte) [inline pkg/operator/deploy/staticresources/aro.openshift.io_clusters.yaml:52]: not sure to see where this is used

**Comment by Makdaam** [comment]
Changing to Ready for Review with the caveat that the text of the message may change.
For now please don't use this feature, message review is in progress in parallel.

File: go.mod
Comment (mjudeikis) [inline go.mod:29]: ?

File: pkg/operator/controllers/banner/banner.go
Comment (mjudeikis) [inline pkg/operator/controllers/banner/banner.go:24]: err lost?

File: pkg/operator/controllers/banner/banner.go
Comment (mjudeikis) [inline pkg/operator/controllers/banner/banner.go:41]: Can we make message above `Sprintf` so we can use here `fmt.Sprint(TextContactSupport, resourceID)` so full message is seen in one place

File: pkg/operator/controllers/banner/banner.go
Comment (mjudeikis) [inline pkg/operator/controllers/banner/banner.go:44]: Picture how this looks?

**Review by mjudeikis — DISMISSED** [review]

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:24]: If the Banner content is set to 0 there's a "not found" error on every reconcile, I want to avoid having a Get -> if exists ->Delete because even then you can get a time of check, time of use race. Also every error returned from the function is logged which is in general a good idea. 

This causes a lot of errors in the log that aren't actual errors.

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go]: Changed to Sprintf.

File: pkg/operator/deploy/staticresources/aro.openshift.io_clusters.yaml
Comment (Makdaam) [reply pkg/operator/deploy/staticresources/aro.openshift.io_clusters.yaml]: Resolved: old make generate.

File: pkg/operator/controllers/banner/banner.go
Comment (m1kola) [reply pkg/operator/controllers/banner/banner.go:25]: You should still handle errors. In this case you probably need logic like:

* Try to delete
  * If delete returns "not found erro" - treat it as not error
  * If delete returns anything else - this is a legitimate error

Here is a similar case:

https://github.com/Azure/ARO-RP/blob/e4ff2a4328ccf77aa600d332364ae8fd4f7b28d9/pkg/operator/controllers/pullsecret/pullsecret_controller.go#L179-L185

File: pkg/operator/controllers/banner/banner.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner.go:22]: Probably makes more sense to use a named return value here: I imagine it will be less error prone as it will be harder overshadow it in some cases.

File: pkg/operator/controllers/banner/banner.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner.go:18]:
```suggestion
	TextContactSupport = "Please ask your cluster administrator to contact Azure or Red Hat support. Your cluster's resource ID: %s"
```
Fro what I've seen Azure docs most commonly use "resource ID".

File: pkg/operator/controllers/banner/banner_controller.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner_controller.go:63]: I think we also need to watch `consolev1.ConsoleNotification` resources with specific name `BannerName` otherwise the following case is possible:

1. We enable a value on `arov1alpha1.Cluster` object
2. Operator reconciles the cluster object and creates the banner object
3. Customer deletes the banner object - at this point operator will not be aware of the fact that the banner got deleted.

Baner will be created again on the next `arov1alpha1.Cluster` reconciliation might mean quite a lot of time without a banner.

**Review by m1kola — CHANGES_REQUESTED** [review]
Few comments/suggestions.

Also I think this also needs unit tests and deserves a simple e2e test where we enable\disable the banner (we expose a banner to the customer so better be safe).

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (Makdaam) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:16]: Actually I picked int here after experiences with the v3 run command Geneva Action. I never remembered what the string should be so I had to go to the code and find it every time.
With 0,1 you can set it quickly, and I agree that if we get to 4 messages it will be problematic. I hope we can do with 1 maybe 2 messages total.

Comment (m1kola) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:16]: @Makdaam how is epereience is going to look like from SRE perspective? Is it going to be:

1. Get cluster object using get kube object geneva action
2. Modify cluster object and update it with put or patch kubernetes object geneva action?

Or are we going to have a separate geneva action for this? If spearate - we can solve the issue of remembering values by providing a dropdown.

File: pkg/operator/controllers/banner/banner_controller.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller.go:63]: Added "Watches".

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:22]: Fixed.

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (Makdaam) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:15]: Changed to Disabled.

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:25]: Ack, added the if statement to nillify the error if it's an "Not Found".

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (m1kola) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:16]: Ah, I see that you are going to create a powershell script. In this case I think it is still better to have a string: I imagine that the script will be something like "enable/disable banner" and we can put relevant values into the script (so we don't have to remember them), but readability will be better when reading with `oc` or geneva actions.

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:49]: Sent via chat since github doesn't allow attachments in comments.
Black text on yellow background.

**Comment by nilsanderselde** [comment]
Is the text in the banner selectable, so they can highlight and copy the resource ID?

Just wanted to check - sometimes GUI designers choose to make UI elements like that not selectable (you could get the value by inspecting, but that is not convenient).

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (nilsanderselde) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go]: If we are going to have multiple preset messages in the future (which I think would be very useful), then having numbers associated with the messages in the RP might get confusing, especially if we discontinue a message. For example, say we have 5 messages and we discontinue #3... then we want to add a new message. Does the new one go in slot 3, or slot 6? We need to keep in mind that people's local scripts are not always up to date, and we don't want to inadvertently display the wrong message in the console.

Strings would be easier. It could still be easy to use with a script. The script could say something like, "Select the message you want to display in the console" and present a list of options:

0: No message
1: First message text
2: Second message text

The SRE could select an option by (non-permanent, displayed only) number, and then it would be posted. In the background, however, it would be explicitly selecting the desired message by name.

File: pkg/operator/controllers/banner/banner_controller.go
Comment (nilsanderselde) [inline pkg/operator/controllers/banner/banner_controller.go:42]:
```suggestion
// Reconcile posts or removes the notification banner
```

File: pkg/operator/controllers/banner/banner_controller.go
Comment (nilsanderselde) [inline pkg/operator/controllers/banner/banner_controller.go:65]:
```suggestion
		Watches(&source.Kind{Type: &consolev1.ConsoleNotification{}}, &handler.EnqueueRequestForObject{}). // watching ConsoleNotifications in case a user edits it
```

**Review by nilsanderselde — COMMENTED** [review]
Comments look nicer with a space after `//`

File: pkg/operator/controllers/banner/banner.go
Comment (bennerv) [inline pkg/operator/controllers/banner/banner.go:38]: Do we have to specify TypeMeta?  I thought it was inferred based on the struct you were creating for the given type.

**Comment by Makdaam** [comment]
> Is the text in the banner selectable, so they can highlight and copy the resource ID?

Yes, the whole notification message is selectable in the browser.

File: pkg/operator/controllers/banner/banner_controller.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller.go:42]: Added spaces after //

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go]: Turns out we don't. Removed.

Comment (m1kola) [reply pkg/operator/controllers/banner/banner.go]: It depends on the marshaler you use: if you use `encoding/json` from standard library - you have to explicitly set it. If you use scheme aware one - then you don't.

**Review by cadenmarchese — COMMENTED** [review]
Maybe a few small changes for consistency with other controllers, but this is otherwise looking great.

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (Makdaam) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go]: Thanks for the points above. Changed to strings.

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go]: pulled out const to const.go

File: pkg/operator/controllers/banner/banner.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner.go:23]: It is easy to break error handling like. We definitely need a unit test.

Also I suggest considering writing an e2e test (separate PR is fine) where we enable a banner, check that the banner object exists, diable it and check that the object is no longer exists.

UPD: Clarification for unit test, I think we should test via `Reconcile` method to test whole flow, not only `reconcileBanner`.

File: pkg/operator/controllers/banner/banner_controller.go
Comment (m1kola) [reply pkg/operator/controllers/banner/banner_controller.go:66]: Nit: I would also move the comment to its own line

```suggestion
		// watching ConsoleNotifications in case a user edits it
		Watches(&source.Kind{Type: &consolev1.ConsoleNotification{}}, &handler.EnqueueRequestForObject{}).
```

**Review by m1kola — CHANGES_REQUESTED** [review]
Looks good overall, but needs tests before we merge

File: pkg/operator/controllers/banner/const.go
Comment (mjudeikis) [inline pkg/operator/controllers/banner/const.go:8]: `Your cluster is experiencing issues. Please contact Azure Red Hat support . Cluster resource ID: %s`

**Comment by mjudeikis** [comment]
Im tempted to change the message to suggested one and merge. We gave enough time for review...

File: pkg/operator/controllers/banner/const.go
Comment (nilsanderselde) [reply pkg/operator/controllers/banner/const.go:8]: I would remove the space between "support" and the period. Also, I like the original phrasing, "Azure or Red Hat support".

**Comment by m1kola** [comment]
@mjudeikis I think it is important here to have good unit test coverage. https://github.com/Azure/ARO-RP/pull/1589#discussion_r673087288

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:23]: Added unit test for Reconcile. In this case it's almost a 1:1 test with reconcileBanner.

File: pkg/operator/controllers/banner/const.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/const.go]: The message text was approved by Product Management. I changed it to the approved message.

File: pkg/operator/controllers/banner/banner_controller.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller.go]: Done and done.

File: pkg/operator/controllers/banner/const.go
Comment (nilsanderselde) [inline pkg/operator/controllers/banner/const.go:9]:
```suggestion
	TextContactSupport = "We have noticed an issue regarding your cluster requiring an action on your part. Please contact support with your cluster resource ID: %s"
```

Run-on sentence :)

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner_controller_test.go:166]: Just printing error is sufficent, in case of failure it will print test name, test case name, line where it failed and the error.

```suggestion
					t.Error(err)
```

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner_controller_test.go:169]: Same as above. This is sufficent:

```suggestion
					t.Error(resultBanner.Spec.Text)
```

Comment (m1kola) [reply pkg/operator/controllers/banner/banner_controller_test.go:169]: And below

**Review by m1kola — COMMENTED** [review]
Test cases revealed a good question (see [the comment below](https://github.com/Azure/ARO-RP/pull/1589#discussion_r685093308)): do we remove the banner if we disable the feature flag or should we keep it (we currently keep it). I would expect the banner to be removed, but I like to hear other opinions on this.

Once we have an agrement on removing/keeping the banner - I think we can merge. Rest of my comments are just nits and I'm happy for them to be addressed in a fillow up PRs so we don't have to hold this one.

File: pkg/operator/controllers/banner/banner.go
Comment (mjudeikis) [inline pkg/operator/controllers/banner/banner.go:23]: Just return here, this is somehow redundant assignment. 
```
// we don't care if the object doesn't exist
return nil
```

File: pkg/operator/controllers/banner/banner.go
Comment (mjudeikis) [inline pkg/operator/controllers/banner/banner.go:41]: All this switch would be bit more readable like this:
```
switch instance.Spec.Banner.Content {
	case arov1alpha1.BannerDisabled:
		err = r.consolecli.ConsoleV1().ConsoleNotifications().Delete(ctx, BannerName, metav1.DeleteOptions{})
		if err != nil && kerrors.IsNotFound(err) {
			// we don't care if the object doesn't exist
			return nil
		}
		return err
	case arov1alpha1.BannerContactSupport:
		_, err := r.consolecli.ConsoleV1().ConsoleNotifications().Get(ctx, BannerName, metav1.GetOptions{})
		if !kerrors.IsNotFound(err) {
			return err
		}
		
		// create or update
		if kerrors.IsNotFound(err) {
			// if the object doesn't exist Create
			_, err = r.consolecli.ConsoleV1().ConsoleNotifications().Create(ctx, r.newBanner(TextContactSupport, instance.Spec.ResourceID), metav1.CreateOptions{})
			return err
		}


		// if there's no errors, object found then update
		_, err = r.consolecli.ConsoleV1().ConsoleNotifications().Update(ctx, r.newBanner(TextContactSupport, instance.Spec.ResourceID), metav1.UpdateOptions{})
		return err
	}
```
Don't yuo think?

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (mjudeikis) [reply pkg/operator/controllers/banner/banner_controller_test.go:64]: Yes, we should :)

**Review by mjudeikis — COMMENTED** [review]
Few nits

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:30]: Done.

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:34]: "Abort" was the wrong phrase, surfacing the error might be better.

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller_test.go:38]: Done.

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller_test.go:64]: I would expect the code to be completely turned off when the feature flag is off. 

When I turn it off, it's off, not trying to clean up its objects because I might want that object to thave a manual override and stay behind - not get reconciled to nil. I know this use case doesn't apply here (1. we don't need manual overrides on a pop up text in a console, 2. if the need ever arises we could use an ConsoleNotification with a different name), but that's the general behaviour I expect from all reconcilers behind a feature: false means the code is not interfering with anything.

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller_test.go:90]: Changed names.

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:23]: done

File: pkg/operator/controllers/banner/banner.go
Comment (Makdaam) [reply pkg/operator/controllers/banner/banner.go:43]: Early escape looks cleaner since we surface all actual error states early (return err) and then take care of other err values.
Modified the switch.

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (m1kola) [reply pkg/operator/controllers/banner/banner_controller_test.go:64]: We discussed it briefly on a call. The idea is that we use feature flag to disable reconciliation, not to disable the banner. It is just a mechanism which we use to turn off the controller if it is broken.

So basically:

* To disable the banner - set content to `BannerDisabled`.
* To disable reconciler - set `ReconcileBanner` to `false`

If we start using feature flag to clean up the banner then:

* We end up with 2 ways to disable the banner (`BannerDisabled` content and `ReconcileBanner` set to `false`)
* Feature flag no longer serves the porpuse of an emergency brake: we still need to run reconciliation and this reconciliation might have a a bug on it's own so we are screwwed and need to deploy a new operator again.

Comment (Makdaam) [reply pkg/operator/controllers/banner/banner_controller_test.go:64]: Also I think this goes into a wider discussion on what we mean by a feature flag in ARO Operator in general. I'm adding a design/definition doc and Architecture call topic to discuss it.

File: pkg/operator/controllers/banner/banner_controller.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner_controller.go:67]: I think it is a good idea to only watch for console notifications with name `BannerName`.

Maybe a follow up PR?

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner_controller_test.go:178]: It is a nit, but I would just printing error beucase on failure it will print test name, test case name, line where it failed and the error.


```suggestion
			if err != nil {
				if tt.expectedErr == nil {
					t.Error(err)
				}
				if !strings.EqualFold(err.Error(), tt.expectedErr.Error()) {
					t.Error(err)
				}
				return
			}
			resultBanner, err := r.consolecli.ConsoleV1().ConsoleNotifications().Get(context.Background(), BannerName, metav1.GetOptions{})
			if tt.expectBanner {
				if err != nil {
					t.Error(err)
				}
				if !strings.EqualFold(resultBanner.Spec.Text, tt.expectedMessage) {
					t.Error(resultBanner.Spec.Text)
				}
			} else {
				if err != nil && !kerrors.IsNotFound(err) {
					t.Error(err)
				}
				if err == nil || !kerrors.IsNotFound(err) || resultBanner != nil {
					t.Errorf("Expected not to get a ConsoleNotification, but it exists")
				}
			}
```

**Review by m1kola — APPROVED** [review]
Few suggestions, but nothing to block this PR: I'm ok for these to be addressed in a small follow up PR or in this one if you prefer.

Note: e2es failed, but these failures do not seem to be related. Maybe a flake?

File: pkg/operator/controllers/banner/banner_controller_test.go
Comment (m1kola) [inline pkg/operator/controllers/banner/banner_controller_test.go:162]: We need to test that:

1. When the unit under test return an error, we need to make sure that this is the error we expet (you handled this)
2. When there is no error we need to check that this is what we expect (you do not check it)

Below is the pattern which we often use (`wantErr` is a `string`).

```suggestion
			if err != nil && err.Error() != tt.wantErr ||
				err == nil && tt.wantErr != "" {
				t.Error(err)
			}
```

**Review by m1kola — APPROVED** [review]
Looks good to me. Let's merge once e2es pass and address feedback (see comments below) in a separate smaller PR.

**Pull request merged**
